README
courrier is a script for read mail on POP or imap servers.

Based on the script and the skin of gmailchecker

Configure server in settings
You can define the max size of messages and a filter to search messages in IMAP server
You can scroll message text with FASTFORWARD (key f for the keyboard) and REWIND (key r)

At the moment you view message only in text mode and html message are convert in text with Beautilfulsoup and html2text "Aaron Swartz (me@aaronsw.com)" modules.
On bottom you have status line with date and attach file name of the selected message
For IMAP server you can define in SKIN your folder and the flag use for select message



changelog.txt
0.2.4
- Change string ID range
- Fix error library path
- Fix translation
0.2.3
- Add specifics fonts for skin courrier
- Fix error in getpopmail
0.2.2
- Fix error ASCII code with debug messages
- Fix error in pop3.stat function
- Test to Isengard and Jarvis Kodi version
0.2.1
- Fix warnings
0.1.3
- Change python version for HELIX and Gotham
0.1.2
- Fix error when no config file with script.notifier
- Remove TAB in code
0.1.1
- Fix bug in translate messages
0.1
- Remove TODO file
- Remove gmail icon file
- Add Autoscroll text in xml skin file
0.0.9
- Cleanning code
- Change name of skin file
0.0.8
- Add filter seach parameter for IMAP
0.0.7
- Complete IMAP support
0.0.6
- First version with partial IMAP support
0.0.4
- change skin WinMail.py
0.0.3
- add scroll for message body
0.0.2
- add quit button
0.0.1
- use a skin xml file for mail window


test_imap.py
#!/usr/bin/python2
# -*- coding: utf-8 -*-

import email
import imaplib
import re
import html2text
import os
from re import compile as Pattern

import email
from email.Parser import Parser as EmailParser
from email.utils import parseaddr
from email.Header import decode_header

USER = 'fce.valeins@libertysurf.fr'
PASSWORD = 'umr5536'

imap = imaplib.IMAP4('imap.aliceadsl.fr')
att_file = ':'
#imap.login(USER, PASSWORD)
imap.login('fce.valeins@libertysurf.fr', 'umr5536')
imap.select('INBOX')
#Search new mail, Filter examples : UNSEEN, ALL, ....
#numEmails = len(imap.search(None, 'UnSeen')[1][0].split())
numEmails = len(imap.search(None, 'ALL')[1][0].split())
#numEmails = len(imap.search(None, '(FROM "samsung.com")' )[1][0].split())
print( ("You have %d emails IMAP" % numEmails) )
i = 0
##Retrieve list of mails
#typ, data = imap.search('UTF-8', SEARCH_PARAM)
#typ, data = imap.search(None, 'UnSeen')
#typ, data = imap.search(None, 'ALL')
#typ, data = imap.search(None, '(FROM "samsung.com")')
#for num in data[0].split():
#    i = i + 1
#    typ, data = imap.fetch(num, '(RFC822)')
#    up = (i*100)/numEmails    #Get mail              Please wait
#    text = data[0][1].strip()
#    parse()
    #print ('TEXT : %s ' % text)
#imap.logout

def parse():
 
 global text
 global att_file

 myemail = email.message_from_string(text)
 p = EmailParser()
 msgobj = p.parsestr(text)
 if msgobj['Subject'] is not None:
    decodefrag = decode_header(msgobj['Subject'])
    subj_fragments = []
    for s , enc in decodefrag:
        if enc:
            s = unicode(s , enc).encode('utf8','replace')
        subj_fragments.append(s)
    subject = ''.join(subj_fragments)
 else:
    subject = None
 if msgobj['Date'] is not None:
    date = msgobj['Date']
 else:
    date = '--'
 Sujet = subject
 realname = parseaddr(msgobj.get('From'))[1]
 print ("FROM : %s SUJET : %s " % (Sujet,realname))
 body = None
 html = None
 #Repertory for attached file(s)
 detach_dir = '/tmp/'
 counter = -1 
 attached_images = []
 for part in msgobj.walk():
    content_disposition = part.get("Content-Disposition", None)
    prog = re.compile('attachment')
    #Retrieve attached files names 
    if prog.search(str(content_disposition)):
        file_att = str(content_disposition)
	#########################################################################
        filename = part.get_filename()
        counter += 1
        att_path = os.path.join(detach_dir, filename)
        pattern = re.compile('png|jpg')
        if pattern.search(str(att_path)):
            #print(("File : %s" % (att_path)))
            fp = open(att_path, 'wb')
            fp.write(part.get_payload(decode=True))
            fp.close()
            attached_images.append(att_path)
            #print(("ATTACHED :%s" % attached_images))
	##########################################################################
	#print(("ATT : %s" % (file_att)))
	#print(part)
        pattern = Pattern(r"\"(.+)\"")
        att_file +=  str(pattern.findall(file_att))
    if part.get_content_type() == "text/plain":
        if body is None:
            body = ""
        try :
            #If no defined charset
            if (part.get_content_charset() is None):
                body +=  part.get_payload(decode=True)
            else:
                body += unicode(
                   part.get_payload(decode=True),
                   part.get_content_charset(),
                   'replace'
                   ).encode('utf8','replace')
            #print ("BODY : %s " % body)
        except Exception:
            body += "Erreur unicode"
            print( "BODY Erreur = %s " % body)
    elif part.get_content_type() == "text/html":
        if html is None:
            html = ""
        try :
            #Test to try fix error unicode 
            #'ascii' codec can't decode byte 0xc5 in position 32: ordinal not in range(128)
            print ("CHARSET = %s " % part.get_content_charset())
            if (part.get_content_charset() is None):
                raw_html = part.get_payload(decode=True)
                #raw_html.upper()
                #raw_html = raw_html.encode('utf-8','replace')
                print("RAW_HTML None 126")
                try :
                  html = html2text.html2text(raw_html)
                except Exception, e:
                  print ("Error 130 : %s" % e)
                  raw_html = raw_html.decode('utf-8','replace')
                  html = html2text.html2text(raw_html)
                  print("RAW_HTML None OK")
            else:
                raw_html = part.get_payload(decode=True)
                charset_msg = part.get_content_charset()
                #unicode(
                #   part.get_payload(decode=True),
                #   part.get_content_charset(),
                #   'ignore'
                #   ).encode('utf8','ignore')
                #print("RAW_HTML 126 : %s" % raw_html)
                html = raw_html.decode(charset_msg)
                html = html2text.html2text(html)
                print ("HTML OK\n")
                #print ("=========================\n%s\n-----------------------\n" % html)
        except Exception, e:
            #html += "Erreur unicode html"
            print( "ERROR HTML = %s " % (str(e)))
            print( "ERROR HTML =============================  ")
            #print ("PART PAYLOAD %s " % part.get_payload(decode=True))
            #raw_html = 	part.get_payload(decode=True)
            #unicode_coded_entities_html = unicode(BeautifulStoneSoup(raw_html,
            #                convertEntities=BeautifulStoneSoup.HTML_ENTITIES))
            #print ('unicode_coded : %s ' % unicode_coded_entities_html)
            #html += unicode_coded_entities_html
            #html = raw_html	
            #print( "==> HTML avant html2text = %s <==" % html )
            #html = html2text.html2text(raw_html)
            #print( "%s" % html )
        #except Exception, e:
         #   html += "Erreur unicode html"
            print( "ERROR HTML = %s " % (str(e)))
            #print( "==> HTML = %s <==" % html )
        realname = parseaddr(msgobj.get('From'))[1]
        Sujet = subject
        description = ' '
        if (body):
            description = str(body)
        else:
            try:
                html = html.encode('ascii','replace')
                description = str(html)
            except Exception, e:
                print( str(e) )
        #Nb of lines msg for scroll text
        nb_lignes = description.count("\n")

        date += att_file
        #print(("attached files : %s" % att_file))
	#Verify if att_path exist
        if 'attached_images' in locals():
           counter = 0
           for name_image in attached_images:
              #print(('attached %s ' % attached_images))
              counter += 1 
              #print(('image%s, IMAGE : %s ' % (counter,name_image)))
    
##Retrieve list of mails
#typ, data = imap.search('UTF-8', SEARCH_PARAM)
#typ, data = imap.search(None, 'UnSeen')
typ, data = imap.search(None, 'ALL')
#typ, data = imap.search(None, '(FROM "samsung.com")')
for num in data[0].split():
    i = i + 1
    typ, data = imap.fetch(num, '(RFC822)')
    up = (i*100)/numEmails    #Get mail              Please wait
    text = data[0][1].strip()
    parse()
    #print ('TEXT : %s ' % text)
imap.logout


resources/lib/MyFont.py
# script constantes
script       = "MyFont.py"
author       = "Ppic, Frost, modify by Senufo"
credits      = "Team XBMC-Passion, http://passion-xbmc.org/"
platform     = "xbmc media center, [LINUX, OS X, WIN32, XBOX]"
date         = "28-12-2016"
version      = "1.1"

#python librairy to add font to the current skin. need to have font_filename.ttf in /resources/fonts/, this script will automatically add it to current skin when called.

import os
import elementtree.ElementTree as ET
import shutil
from traceback import print_exc
import xbmc
import xbmcaddon

scriptid   = "script.courrier"
DEBUG_LOG = -1

Addon      = xbmcaddon.Addon(scriptid)
scriptpath = Addon.getAddonInfo('path').decode('utf-8')

skin_font_path = xbmc.translatePath("special://skin/fonts/")
script_font_path = os.path.join(scriptpath , "resources" , "skins", "fonts")
skin_dir = xbmc.translatePath("special://skin/")
list_dir = os.listdir( skin_dir )

#xbmc.log(("skin_font_path : %s, script_font_path : %s, skindir : %s, list_dir : %s" % (skin_font_path, script_font_path, skin_dir,list_dir)), , DEBUG_LOG)

def getFontsXML():
    fontxml_paths = []
    try:
        for item in list_dir:
            item = os.path.join( skin_dir, item )
            if os.path.isdir( item ):
                font_xml = os.path.join( item, "Font.xml" )
                if os.path.exists( font_xml ):
                    fontxml_paths.append( font_xml )
    except:
        print_exc()
    return fontxml_paths

def isfontinstalled( fontxml_path, fontname ):
    name = "<name>%s</name>" % fontname
    xbmc.log(("[%s] : font name in %s : %s" % (scriptid, fontxml_path, name)), DEBUG_LOG)
    if not name in file( fontxml_path, "r" ).read():
        xbmc.log(("[%s] : font name not installed! : %s" % (scriptid, fontname)), DEBUG_LOG)
        return False
    else:
        xbmc.log(("[%s] : ok install font name : %s" % (scriptid, fontname)), DEBUG_LOG)
        return True

def addfont( fontname , filename , size , style="", aspect="" ):
    try:
        reload_skin = False
        fontxml_paths = getFontsXML()

        if fontxml_paths:
            for fontxml_path in fontxml_paths:
                xbmc.log(("[%s] : analyse du fichier: %s" % (scriptid, fontxml_path)), DEBUG_LOG)
                if not isfontinstalled( fontxml_path, fontname ):
                    tree = ET.parse(fontxml_path)
                    root = tree.getroot()
                    xbmc.log(("[%s] : modification du fichier: %s" % (scriptid, fontxml_path)), DEBUG_LOG)
                    for sets in root.getchildren():
                        sets.findall( "font" )[ -1 ].tail = "\n\t\t" #"\n\n\t\t"
                        new = ET.SubElement(sets, "font")
                        new.text, new.tail = "\n\t\t\t", "\n\t"
                        subnew1=ET.SubElement(new ,"name")
                        subnew1.text = fontname
                        subnew1.tail = "\n\t\t\t"
                        subnew2=ET.SubElement(new ,"filename")
                        subnew2.text = ( filename, "Arial.ttf" )[ sets.attrib.get( "id" ) == "Arial" ]
                        subnew2.tail = "\n\t\t\t"
                        subnew3=ET.SubElement(new ,"size")
                        subnew3.text = size
                        subnew3.tail = "\n\t\t\t"
                        last_elem = subnew3
                        if style in [ "normal", "bold", "italics", "bolditalics" ]:
                            subnew4=ET.SubElement(new ,"style")
                            subnew4.text = style
                            subnew4.tail = "\n\t\t\t"
                            last_elem = subnew4
                        if aspect:
                            subnew5=ET.SubElement(new ,"aspect")
                            subnew5.text = aspect
                            subnew5.tail = "\n\t\t\t"
                            last_elem = subnew5
                        reload_skin = True

                        last_elem.tail = "\n\t\t"
                    tree.write(fontxml_path)
                    reload_skin = True
    except:
        print_exc()

    if reload_skin:
        if not os.path.exists( os.path.join( skin_font_path, filename ) ) and os.path.exists( os.path.join( script_font_path, filename ) ):
            shutil.copyfile( os.path.join( script_font_path, filename ), os.path.join( skin_font_path, filename ) )

        xbmc.executebuiltin( "XBMC.ReloadSkin()" )
        return True

    return False


addon.py
# -*- coding: utf-8 -*-

"""
kodi script for read mail on IMAP/POP server.
"""
# Modules xbmc
import xbmc
import xbmcgui
import xbmcaddon
import os, re, glob
from re import compile as Pattern
import sys
import time
import poplib, imaplib
import string

import email
from email.Parser import Parser as EmailParser
from email.utils import parseaddr
from email.Header import decode_header

author     = "Senufo"
scriptid   = "script.courrier"
scriptname = "Courrier"

Addon      = xbmcaddon.Addon(scriptid)

version    = Addon.getAddonInfo('version')
language   = Addon.getLocalizedString
scriptpath = Addon.getAddonInfo('path').decode('utf-8')

skindir    = xbmc.getSkinDir()

DEBUG_LOG = Addon.getSetting('Debug')
if 'true' in DEBUG_LOG: DEBUG_LOG = -1 #loglevel == 1 (DEBUG, shows all)
else: DEBUG_LOG = 1 #(NONE, nothing at all is logged)
xbmc.log(("[%s] : DEBUG_LOG : %s" % (scriptid,DEBUG_LOG)),DEBUG_LOG)
xbmc.log(("[%s] : skindir : %s, scriptpath : %s" % (scriptid,skindir,scriptpath)),DEBUG_LOG)# DEBUG_LOG = True


#Add path for my library
sys.path.append (xbmc.translatePath( os.path.join( scriptpath, 'resources', 'lib' ) ))
# Script html2text.py in resources/lib
import html2text
#for add a specific font to the selectd skin
import MyFont
MyFont.addfont( "Font24","aeon_common.ttf", "24")
MyFont.addfont( "Font_titlemail","aeon_mainmenu1.ttf","40")
MyFont.addfont( "Font_datemail","aeon_mainmenu1.ttf","30")
MyFont.addfont( "Font_buttonmail","aeon_mainmenu6.ttf","25")
MyFont.addfont( "Font_msg","aeon_mainmenu8.ttf","30")

# Defaults options for html2text module
UNICODE_SNOB = Addon.getSetting('UNICODE_SNOB')                  # UNICODE_SNOB=0
ESCAPE_SNOB = Addon.getSetting('ESCAPE_SNOB')                    # ESCAPE_SNOB=0
LINKS_EACH_PARAGRAPH = Addon.getSetting('LINKS_EACH_PARAGRAPH')  # LINKS_EACH_PARAGRAPH=0
BODY_WIDTH = Addon.getSetting('BODY_WIDTH')                      # BODY_WIDTH=78
SKIP_INTERNAL_LINKS = Addon.getSetting('SKIP_INTERNAL_LINKS')    # SKIP_INTERNAL_LINKS=True
if 'true' in SKIP_INTERNAL_LINKS: SKIP_INTERNAL_LINKS = True
else: SKIP_INTERNAL_LINKS = False
INLINE_LINKS = Addon.getSetting('INLINE_LINKS')                  # INLINE_LINKS=True
if 'true' in INLINE_LINKS: INLINE_LINKS = True
else: INLINE_LINKS = False
GOOGLE_LIST_INDENT = Addon.getSetting('GOOGLE_LIST_INDENT')      # GOOGLE_LIST_INDENT=36
IGNORE_ANCHORS = Addon.getSetting('IGNORE_ANCHORS')              # IGNORE_ANCHORS=False
if 'true' in IGNORE_ANCHORS: IGNORE_ANCHORS = True
else: IGNORE_ANCHORS = False
IGNORE_IMAGES = Addon.getSetting('IGNORE_IMAGES')                # IGNORE_IMAGES=True
if 'true' in IGNORE_IMAGES: IGNORE_IMAGES = True
else: IGNORE_IMAGES = False
IGNORE_EMPHASIS = Addon.getSetting('IGNORE_EMPHASIS')            # IGNORE_EMPHASIS=False
if 'true' in IGNORE_EMPHASIS: IGNORE_EMPHASIS = True
else: IGNORE_EMPHASIS = False


# Directory for attached file(s)
# Test if directory for attached file exist
try:
    tmp = "%s/%s" % (scriptpath, 'tmp')
    xbmc.log(("[%s] : Directory of attached files : %s " % (scriptid, tmp)), DEBUG_LOG)
    DATA_PATH = xbmc.translatePath(tmp)
except Exception, e:
    xbmc.log(("[%s] : attached files error : %s" % (scriptid, e)), DEBUG_LOG)
if not os.path.exists(DATA_PATH):
    os.makedirs(DATA_PATH)
# if directory exist remove all files
else:
    files = glob.glob("%s/%s" % (DATA_PATH, '*'))
    for f in files:
        try:
            os.remove(f)
            xbmc.log(("[%s] : file : %s" % (scriptid, f)), DEBUG_LOG)
        except:
            xbmc.log(('[%s] : no file' % scriptid), DEBUG_LOG)

#Addon = xbmcaddon.Addon(scriptid)
# get actioncodes from keymap.xml/ keys.h
ACTION_PREVIOUS_MENU = 10
ACTION_SELECT_ITEM     = 7
ACTION_MOVE_LEFT       = 1
ACTION_MOVE_RIGHT      = 2
ACTION_MOVE_UP         = 3
ACTION_MOVE_DOWN       = 4
ACTION_PAGE_UP         = 5
ACTION_PAGE_DOWN       = 6
ACTION_NUMBER1         = 59
ACTION_NUMBER2         = 60
ACTION_VOLUME_UP       = 88
ACTION_VOLUME_DOWN     = 89
ACTION_REWIND          = 78
ACTION_FASTFORWARD     = 77

# Buttons ID in script-courrier-main.xml
STATUS_LABEL = 100
NX_MAIL      = 101
MSG_BODY     = 102
EMAIL_LIST   = 120
SCROLL_BAR   = 121
MSG_BODY     = 102
SERVER1      = 1001
SERVER2      = 1002
SERVER3      = 1003
QUIT         = 1004
FILE_ATT     = 1005
MAX_SIZE_MSG = int(Addon.getSetting('max_msg_size'))
SEARCH_PARAM = Addon.getSetting('search_param')
BACKGROUND   = Addon.getSetting('background')

# Test if BACKGROUND is defined
if not BACKGROUND:
    BACKGROUND = 'SKINDEFAULT.jpg'


class MailWindow(xbmcgui.WindowXML):
    """
    Display main window for read mail.
    """
    def __init__(self, *args, **kwargs):
        """
        init position variable.
        """
        # variable for position in the msg
        self.position = 0

    def onInit(self):
        """
        Initialize parameters.
        """
        self.getControl(EMAIL_LIST).reset()
        for i in [1, 2, 3]:
            id = 'name' + str(i)
            NOM = Addon.getSetting(id)
            Button_Name = 1000 + i
            if NOM:
                self.getControl(Button_Name).setLabel(NOM)
            else:
                self.getControl(Button_Name).setEnabled(False)
        self.checkEmail(Addon.getSetting('name1'))


# Verify mails and display subjects and expeditors
# Alias is name server POP or IMAP
    def checkEmail(self, alias):
        """
        Check mail on POP or IMAP server.
        """
        # debug( 'ALIAS = %s ' % alias )
        self.getControl(STATUS_LABEL).setLabel('%s ...' % alias)

        # Empty the list of subject messages
        #self.listControl.reset()
        self.USER = ''
        self.NOM = ''
        self.SERVER = ''
        self.PASSWORD = ''
        self.PORT = ''
        self.SSL = ''
        self.TYPE = ''
        self.FOLDER = ''
        # Get list of the 3 servers in settings.xml
        for i in [1, 2, 3]:
            NOM = Addon.getSetting('name%i' % i)
            USER = Addon.getSetting('user%i' % i)
            NOM = Addon.getSetting('name%i' % i)
            SERVER = Addon.getSetting('server%i' % i)
            PASSWORD = Addon.getSetting('pass%i' % i)
            PORT = Addon.getSetting('port%i' % i)
            SSL = Addon.getSetting('ssl%i' % i) == "true"
            TYPE = Addon.getSetting('type%i' % i)
            FOLDER = Addon.getSetting('folder%i' % i)

            # Search selected server
            if (alias == NOM):
                self.NOM = NOM
                self.SERVER = SERVER
                self.USER = USER
                self.PORT = PORT
                self.PASSWORD = PASSWORD
                self.TYPE = TYPE
                self.SSL = SSL
                self.FOLDER = FOLDER
                try:
                    # Select server type
                    if '0' in self.TYPE:  # POP3
                        self.getPopMails()
                    if '1' in self.TYPE:  # IMAP
                        self.getImapMails()
                except Exception, e:
                    xbmc.log(('[%s] : Error Server Type : %s' % (scriptid, str(e))), DEBUG_LOG)
                    dialog = xbmcgui.DialogProgress()
                    dialog.create(Addon.getLocalizedString(id=614),
                                  Addon.getLocalizedString(id=620) % self.SERVER)
                    time.sleep(5)
                    dialog.close()

    def getPopMails(self):
        """
        Get mail on POP server with poplib.
        """
        dialog = xbmcgui.DialogProgress()
        dialog.create(Addon.getLocalizedString(id=614),
                      Addon.getLocalizedString(id=610))  # Inbox, Logging in
        if self.SSL:
            mail = poplib.POP3_SSL(str(self.SERVER), int(self.PORT))
        else:  # 'POP3'
            mail = poplib.POP3(str(self.SERVER), int(self.PORT))
        mail.user(str(self.USER))
        mail.pass_(str(self.PASSWORD))
        try:
            numEmails = mail.stat()[0]
        except Exception, e:
            xbmc.log(("[%s] : Erreur stat : %s<=" % (scriptid,str(e))), DEBUG_LOG)

        xbmc.log(("[%s] : You have %d emails " % (scriptid, numEmails)), DEBUG_LOG)
        # Display the number of msg
        self.getControl(NX_MAIL).setLabel('%d msg(s)' % numEmails)
        dialog.close()
        if numEmails == 0:
            dialogOK = xbmcgui.Dialog()
            dialogOK.ok("%s" % self.NOM,
                        Addon.getLocalizedString(id=612))  # no mail
            self.getControl(EMAIL_LIST).reset()
        else:
            dialog.create(Addon.getLocalizedString(id=613),  # Inbox
                          Addon.getLocalizedString(id=615) + str(numEmails) + # You have
                          Addon.getLocalizedString(id=616)) # emails
            # Retrieve list of mails
            resp, items, octets = mail.list()
            xbmc.log(("[%s] : resp %s " % (scriptid, resp)), DEBUG_LOG)
            xbmc.log(("[%s] : items %s" % (scriptid, items)), DEBUG_LOG)
            dialog.close()
            # Get all messages for display
            progressDialog = xbmcgui.DialogProgress()
            progressDialog.create(Addon.getLocalizedString(id=617),  # Message(s)
                                  Addon.getLocalizedString(id=618))  # Get mail
            i = 0
            # Reset ListBox msg
            self.getControl(EMAIL_LIST).reset()
            self.emails = []
            for item in items:
                i = i + 1
                id, size = string.split(item)
                up = (i*100)/numEmails
                progressDialog.update(up,
                                      Addon.getLocalizedString(id=618),  # Get mail
                                      Addon.getLocalizedString(id=619))  # Please wait

                # If the maximum size is exceeded doxnload only 50 lines
                if (MAX_SIZE_MSG == 0) or (size < MAX_SIZE_MSG):
                    resp, text, octets = mail.retr(id)
                else:
                    resp, text, octets = mail.top(id, 300)
                att_file = ':'
                text = string.join(text, "\n")
                self.processMails(text, att_file)
            progressDialog.close()
            # Display the first mail of the list
            self.getControl(EMAIL_LIST).selectItem(0)

    def processMails(self, text, att_file):
        """
        Parse mail for display in XBMC.
        """
        myemail = email.message_from_string(text)
        p = EmailParser()
        msgobj = p.parsestr(text)
        if msgobj['Subject'] is not None:
            decodefrag = decode_header(msgobj['Subject'])
            subj_fragments = []
            for s, enc in decodefrag:
                # Encode subject in UTF-8
                if enc:
                    s = unicode(s, enc).encode('utf8', 'replace')
                subj_fragments.append(s)
            subject = ''.join(subj_fragments)
        else:
            subject = None
        date = Addon.getLocalizedString(32132)
        if msgobj['Date'] is not None:
            date += msgobj['Date']
        Sujet = subject
        realname = parseaddr(msgobj.get('From'))[1]
        xbmc.log(("[%s] : SUJET : %s " % (scriptid, Sujet)), DEBUG_LOG)
        body = None
        html = None
        counter = -1
        attached_images = []
        for part in msgobj.walk():
            content_disposition = part.get("Content-Disposition", None)
            prog = re.compile('attachment')
            # Retrieve attached files names
            if prog.search(str(content_disposition)):
                file_att = str(content_disposition)
                filename = part.get_filename()
                counter += 1
                att_path = os.path.join(DATA_PATH, filename)
                pattern = re.compile('png|jpg|gif')
                if pattern.search(str(att_path)):
                    xbmc.log(("[%s] : File : %s" % (scriptid, att_path)), DEBUG_LOG)
                    fp = open(att_path, 'wb')
                    fp.write(part.get_payload(decode=True))
                    fp.close()
                    attached_images.append(att_path)
                    xbmc.log(("[%s] : ATTACHED :%s" % (scriptid, attached_images)), DEBUG_LOG)
                pattern = Pattern(r"\"(.+)\"")
                att_file += str(pattern.findall(file_att))
            # Treat text/plain msg
            if part.get_content_type() == "text/plain":
                if body is None:
                    body = ""
                    try:
                        # If no defined charset
                        if (part.get_content_charset() is None):
                            # Decode charset and encode in UTF-8
                            body += unicode(part.get_payload(decode=True)).encode('utf8', 'replace')
                        else:
                            # Decode charset and encode in UTF-8
                            body += unicode(
                                part.get_payload(decode=True),
                                part.get_content_charset(),
                                'replace'
                                ).encode('utf8', 'replace')
                    except Exception, e:
                        # body += "Erreur unicode : %s" % e
                        body += part.get_payload(decode=True)
                    #xbmc.log(("[%s] : 369: BODY = %s " % (scriptid, body)), DEBUG_LOG)
            elif part.get_content_type() == "text/html":
                # Define defaults parameters for html2text object
                h = html2text.HTML2Text()
                h.unicode_snob = int(UNICODE_SNOB)                  # UNICODE_SNOB=0
                h.escape_snob = int(ESCAPE_SNOB)                    # ESCAPE_SNOB=0
                h.links_each_paragraph = int(LINKS_EACH_PARAGRAPH)  # LINKS_EACH_PARAGRAPH=0
                h.body_width = int(BODY_WIDTH)                      # BODY_WIDTH=78
                h.skip_internal_links = bool(SKIP_INTERNAL_LINKS)   # SKIP_INTERNAL_LINKS=True
                h.inline_links = bool(INLINE_LINKS)                 # INLINE_LINKS=True
                h.google_list_indent = int(GOOGLE_LIST_INDENT)      # GOOGLE_LIST_INDENT=36
                h.ignore_links = bool(IGNORE_ANCHORS)               # IGNORE_ANCHORS=False
                h.ignore_images = bool(IGNORE_IMAGES)               # IGNORE_IMAGES=True
                h.ignore_emphasis = bool(IGNORE_EMPHASIS)           # IGNORE_EMPHASIS=False

                if html is None:
                    html = ""
                try:
                    # Test to try fix error unicode
                    # 'ascii' codec can't decode byte 0xc5 in position 32: ordinal not in range(128)
                    # debug ("CHARSET = %s " % part.get_content_charset())
                    if (part.get_content_charset() is None):
                        raw_html = part.get_payload(decode=True)
                        try:
                            html = h.handle(raw_html)
                        # Try to fix error unicode if no charset defined
                        except Exception, e:
                            xbmc.log(("[%s] : Error : %s, CHARSET = %s " % (scriptid, e, part.get_content_charset())), DEBUG_LOG)
                            raw_html = raw_html.decode('utf-8', 'replace')
                            html = h.handle(raw_html)
                            xbmc.log(("[%s] : RAW_HTML None OK" % scriptid), DEBUG_LOG)
                    else:
                        # Decode in UTF-8 with kown charset
                        raw_html = part.get_payload(decode=True)
                        charset_msg = part.get_content_charset()
                        xbmc.log(("[%s] : CHARSET text/html = %s " % (scriptid, part.get_content_charset())), DEBUG_LOG)
                        try:
                            html = raw_html.decode(charset_msg)
                            html = h.handle(html)
                        except Exception, e:
                            xbmc.log(("[%s] : HTML error : %s\n" % (scriptid, e)), DEBUG_LOG)
                            # html = htlm2text(raw_html)
                            xbmc.log(("[%s] : HTML OK : %s\n" % (scriptid, html)), DEBUG_LOG)
                except Exception, e:
                    xbmc.log(("[%s] : ERROR HTML = %s , %s" % (scriptid, str(e), charset_msg)), DEBUG_LOG)

            realname = parseaddr(msgobj.get('From'))[1]
        Sujet = subject
        description = '*'
        # if text/plain Body is define
        if (body):
            try:
                description += body.encode('utf-8', 'replace')
            except Exception, e:
                xbmc.log(("[%s] : Error Body text/plain : %s " % (scriptid, e)), DEBUG_LOG)
                description += body.decode('utf-8', 'replace')
        # if text/html html is define
        if (html):
            try:
                # debug('Charset : %s ' % charset_msg)
                description += html.encode('utf-8', 'replace')
            except Exception, e:
                xbmc.log(("[%s] : DESC error : %s" % (scriptid, str(e))), DEBUG_LOG)
        # Nb of lines msg for scroll text
        self.nb_lignes = description.count("\n")
        listitem = xbmcgui.ListItem(label2=realname, label=Sujet)
        listitem.setProperty("realname", realname)
        att_file = Addon.getLocalizedString(32133) + att_file
	try:
	    xbmc.log(("[%s] : attached files : %s" % (scriptid, att_file)), DEBUG_LOG)
	except:
	    xbmc.log(("[%s] : Erreur code ASCII" % scriptid), DEBUG_LOG)
        listitem.setProperty("date", date)
        listitem.setProperty("att_files", att_file)
        description = description + '*'
        listitem.setProperty("message", description)
    # Verify if att_path exist
        if 'attached_images' in locals():
            counter = 0
            for name_image in attached_images:
                xbmc.log(('[%s] : attached %s ' % (scriptid, attached_images)), DEBUG_LOG)
                counter += 1
                xbmc.log(('[%s] : image%s, IMAGE : %s ' % (scriptid, counter, name_image)), DEBUG_LOG)
                listitem.setProperty(('image%s' % counter), name_image)

        # Add the background and color
        listitem.setProperty('background', BACKGROUND)
        listitem.setProperty('couleur', 'yellow')

        self.getControl(EMAIL_LIST).addItem(listitem)

    def getImapMails(self):
        """
        Get mail form IMAP server.
        """
        dialog = xbmcgui.DialogProgress()
        dialog.create(Addon.getLocalizedString(id=614),
                      Addon.getLocalizedString(id=610))  # Inbox,Logging in
        # Reset ListBox msg
        self.getControl( EMAIL_LIST ).reset()
        self.emails = []
        try:
            if self.SSL:
                imap = imaplib.IMAP4_SSL(str(self.SERVER), int(self.PORT))
            else:
                imap = imaplib.IMAP4(str(self.SERVER), int(self.PORT))
            att_file = ':'
            imap.login(self.USER, self.PASSWORD)
            imap.select(self.FOLDER)
        # Search new mail, Filter examples : UNSEEN, ALL, ....
            # numEmails = len(imap.search(None, 'UnSeen')[1][0].split())
            # numEmails = len(imap.search(None, 'ALL')[1][0].split())
            numEmails = len(imap.search(None, SEARCH_PARAM)[1][0].split())
            xbmc.log(("[%s] : You have %d emails IMAP" % (scriptid,numEmails)), DEBUG_LOG)
            # Display number of msg
            self.getControl(NX_MAIL).setLabel('%d msg(s)' % numEmails)
            dialog.close()
            if numEmails == 0:
                dialogOK = xbmcgui.Dialog()
                dialogOK.ok("%s" % self.NOM,
                            Addon.getLocalizedString(id=612))  # no mail
                self.getControl(EMAIL_LIST).reset()
            else:
                progressDialog2 = xbmcgui.DialogProgress()
                progressDialog2.create(Addon.getLocalizedString(id=617),  # Message(s)
                                       Addon.getLocalizedString(id=618))  # Get mail
                i = 0
#        ##Retrieve list of mails
                typ, data = imap.search('UTF-8', SEARCH_PARAM)
                # typ, data = imap.search(None, 'UnSeen')
                # typ, data = imap.search(None, 'ALL')
                for num in data[0].split():
                    i = i + 1
                    typ, data = imap.fetch(num, '(RFC822)')
                    up = (i*100)/numEmails
                    progressDialog2.update(up,
                                           Addon.getLocalizedString(id=618),  # Get mail
                                           Addon.getLocalizedString(id=619))  # Please wait
                    xbmc.log(("[%s] : UP = %d " % (scriptid, up)), DEBUG_LOG)
                    text = data[0][1].strip()
                    self.processMails(text, att_file)
                progressDialog2.close()
            # Display first mail of the list
            self.getControl(EMAIL_LIST).selectItem(0)
            imap.logout
        except Exception, e:
            xbmc.log(('[%s] : IMAP exception : %s' % (scriptid, str(e))), DEBUG_LOG)

    def onAction(self, action):
        """
        Select action after remote control.
        """
        if action == ACTION_PREVIOUS_MENU:
            self.close()
        if action == ACTION_MOVE_UP:
            controlId = action.getId()
        if action == ACTION_MOVE_DOWN:
            controlId = action.getButtonCode()
        if action == ACTION_FASTFORWARD:  # PageUp
            if (self.position > 0):
                self.position = self.position - 1
            self.getControl(MSG_BODY).scroll(self.position)
            xbmc.log(("[%s] : Position F = %d " % (scriptid, self.position)), DEBUG_LOG)
        if (action == ACTION_REWIND):  # PageUp
            if (self.position <= self.nb_lignes):
                self.position = self.position + 1
            self.getControl(MSG_BODY).scroll(self.position)
            xbmc.log(("[%s] : Position R = %d " % (scriptid, self.position)), DEBUG_LOG)

    def onClick(self, controlId):
        """Action on click button"""
        if (controlId in [SERVER1, SERVER2, SERVER3]):
            label = self.getControl(controlId).getLabel()
            self.checkEmail(label)
        elif (controlId == QUIT):
            self.close()

#Create window for display mails
defaultSkin = 'Default'
xbmc.log(('[%s] : %s, %s' % (scriptid,scriptpath, defaultSkin)), DEBUG_LOG)
mydisplay = MailWindow("script-courrier-main.xml", scriptpath, defaultSkin)
mydisplay .doModal()
del mydisplay
